// @mattrglobal/node-bbs-signatures/lib/bls12381.js
"use strict";
/*
 * Copyright 2020 - MATTR Limited
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateBlindedBls12381G2KeyPair = exports.generateBls12381G2KeyPair = exports.generateBlindedBls12381G1KeyPair = exports.generateBls12381G1KeyPair = void 0;
/**
 * @ignore
 */
// eslint-disable-next-line @typescript-eslint/no-var-requires
var path = require("path");
// eslint-disable-next-line @typescript-eslint/no-var-requires
var bbs = require(path.resolve(path.join(__dirname, "../native/index.node")));
/**
 * Generates a BLS12-381 key pair where the public key is a commitment in G1
 * @param seed [Optional] To derive the key pair from
 *
 * @returns A BlsKeyPair
 */
exports.generateBls12381G1KeyPair = function (seed) {
    var result = seed ? bbs.bls_generate_g1_key(seed === null || seed === void 0 ? void 0 : seed.buffer) : bbs.bls_generate_g1_key();
    return {
        publicKey: new Uint8Array(result.publicKey),
        secretKey: new Uint8Array(result.secretKey),
    };
};
/**
 * Generates a blinded BLS12-381 key pair where the public key is a commitment in G1 to the private key
 * along with a further commitment of a blinding factor to the blinding factor generator point in G1
 * @param seed [Optional] To derive the key pair from
 *
 * @returns A BlindedBlsKeyPair
 */
exports.generateBlindedBls12381G1KeyPair = function (seed) {
    var result = seed ? bbs.bls_generate_blinded_g1_key(seed === null || seed === void 0 ? void 0 : seed.buffer) : bbs.bls_generate_blinded_g1_key();
    return {
        publicKey: new Uint8Array(result.publicKey),
        secretKey: new Uint8Array(result.secretKey),
        blindingFactor: new Uint8Array(result.blindingFactor),
    };
};
/**
 * Generates a BLS12-381 key pair where the public key is a commitment in G2
 * @param seed [Optional] To derive the key pair from
 *
 * @returns A BlsKeyPair
 */
exports.generateBls12381G2KeyPair = function (seed) {
    var result = seed ? bbs.bls_generate_g2_key(seed === null || seed === void 0 ? void 0 : seed.buffer) : bbs.bls_generate_g2_key();
    return {
        publicKey: new Uint8Array(result.publicKey),
        secretKey: new Uint8Array(result.secretKey),
    };
};
/**
 * Generates a blinded BLS12-381 key pair where the public key is a commitment in G2 to the private key
 * along with a further commitment of a blinding factor to the blinding factor generator point in G2
 * @param seed [Optional] To derive the key pair from
 *
 * @returns A BlindedBlsKeyPair
 */
exports.generateBlindedBls12381G2KeyPair = function (seed) {
    var result = seed ? bbs.bls_generate_blinded_g2_key(seed === null || seed === void 0 ? void 0 : seed.buffer) : bbs.bls_generate_blinded_g2_key();
    return {
        publicKey: new Uint8Array(result.publicKey),
        secretKey: new Uint8Array(result.secretKey),
        blindingFactor: new Uint8Array(result.blindingFactor),
    };
};
