// jsonld-lint/lib/utilities.js
"use strict";
/*
 * Copyright 2020 - MATTR Limited
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var types_1 = require("./types");
/**
 * Converts a document offset to a position
 * @param offset positional offset of the element
 * @param length length of the element
 */
exports.documentOffSetToPosition = function (offset, length) {
    var startPositionOffset = offset;
    var endPositionOffset = (offset + length);
    return {
        startPositionOffset: startPositionOffset,
        endPositionOffset: endPositionOffset
    };
};
/**
 * Creates a JSON-LD lint error
 * @param type Type of error to create
 * @param message Error message
 * @param details Error details
 */
exports.createJsonLdLintError = function (type, message, details) {
    return {
        name: types_1.JsonLdLintErrorName,
        type: type,
        message: message,
        details: details
    };
};
